### DISCUSSION TITLE
Template commit_message

### DISCUSSION THREAD

**Discussion by mercurykd** [timbrel/GitSavvy#1538] [Ideas]
I saw that you deactivated the pre-commit-message hook. Can it be possible to implement the functionality of editing the commit_message template?
For example, I want to automatically add the branch name to the beginning of the commit.

**Comment by kaste** [comment]
I'm not sure I understand what you mean.  What do you mean with "commit_message template"?

**Comment by mercurykd** [comment]
![image](https://user-images.githubusercontent.com/30900414/161743852-eb688dc6-05da-4399-9ed0-d75aea49e489.png)

**Comment by kaste** [comment]
There is something like this but static

https://github.com/timbrel/GitSavvy/blob/6452528d22a13c6f3169ad20619471d261e576bf/GitSavvy.sublime-settings#L129-L137

T.i. you can create a file which will be read and displayed here.  But it is static content and *below* the help text.

This is generally done here https://github.com/timbrel/GitSavvy/blob/6452528d22a13c6f3169ad20619471d261e576bf/core/commands/commit.py#L121

although I don't know how the feature would look like tbh.

Should this really be automatic *every* time you open the commit message view?  Wouldn't it be preferable to make a shortcut (`ctrl+whatever`) which will then insert the information, e.g. the branch name, at your will?  Or a shortcut which opens a quick panel/menu and offers various information bits you might want to use and where the user can choose from?  

Or make a different command/binding and instead of calling just `gs_commit`, call

```
    {
        "caption": "git: commit with some initial text",
        "command": "gs_commit",
        "args": { "initial_text": "[galore] "}
    },
```

**Reply by IPWright83** [reply]
I think there are some advantages to this being automatic. We having coding standards where it's required, so it's nice if it were just there, but definitely feels like an optional feature.

**Comment by kaste** [comment]
Besides that, at the time we disabled the pre-commit-message hook feature, git core was already working on making a "hooks" API.  As soon as this feature lands, it is in the 2.36rc atm, we can probably look into enabling the hook again.

**Comment by mercurykd** [comment]
It would be cool to make a file that would insert text before COMMIT_HELP_TEXT. With the ability to replace the #branch# template with the name of a branch, for example. Implement your prepare-commit-msg with environment variables.

**Comment by kaste** [comment]
Well, there are some interesting bits here but I think overall a first implementation would be *very* idiosyncratic and of course proprietary.  I think you should try an implementation as an add-on plugin to GitSavvy.  (Actually I start a lot of the functionality as add-ons and some of them land in GitSavvy after months or years, and of course some never get useful and interesting enough.)  

The name of such add-ons must start with GitSavvy, e.g. `GitSavvy-addon-message-templates`, otherwise you can't import anything from `GitSavvy` itself, t.i. you can't use the code form it.  

The command of interest here is `gs_commit`.  Either you write a new command and bind it to the Command Palette and some keys which does some work and then calls `gs_commit` in turn with `initial_text` set.  Just like a normal functional wrapper.  Or, and this is maybe better, you intercept `gs_commit` invocations and rewrite them, also computing an `initial_text` if none is given.

**Reply by mercurykd** [reply]
this is a wish to make a feature. I don't understand the innards of sublime text and I can't implement what you said

**Reply by kaste** [reply]
Sure it's a feature request, but currently I don't see a good fit *within* GitSavvy core for it.  I need something to look at, an implementation, to consider adding it to core.

**Reply by IPWright83** [reply]
Do you have an example of an addon @kaste? I've very little python experience and don't know the Sublime internals either, but happy to give it a shot. I'm thinking of something like:

- Read git branch name
- Allow a regex match on that (so you can support matching groups)
- Specify a template that can consume those matching groups
- Spit the commit message as a prefix into the GitSavvy commit screen

**Reply by kaste** [reply]
I don't have a plugin/addon available/published.  But you basically create a folder prefixed `GitSavvy-` so it is loaded *after* GitSavvy.  

You either make a new command or intercept GitSavvy's own command (and rewrite it).  That's probably more confusing so make a new command as a starter.  Within that file create a file, name doesn't matter, like `plugin.py`:

```
from GitSavvy.core.git_command import GitCommand
from sublime_plugin import WindowCommand
class prepare_commit_message(WindowCommand, GitCommand):
	def run(self):  
		# you now have access to git commands
		print(self.get_current_branch_name())
		# and you can also execute arbitrary git commands
		output = self.git("log", "-1")
		print(output)
 		# do whatever you want here, finally call
		window.run_command("gs_commit", {"initial_text": <here comes the commit message>})
		# to open the commit view
		
```

In the plugin folder you create a file `Default.sublime-keymap` and bind your command

```
[
	{ "keys": ["ctrl+shift+c"], "command": "prepare_commit_message"}
]
```

Usually you try stuff in Sublime's console - REPL driven development.  Just be sure to switch to the python3.3 repl because GitSavvy is still on 33.  (Basic python stuff you develop in ipython which is *the* convenient python REPL.)

**Reply by kaste** [reply]
It's not rocket science, really
